File: .skills/openclaw-skills/skills/purch-agent/agentic-commerce/scripts/buy_and_sign.py
#!/usr/bin/env python3
"""
Create a purchase order AND sign/submit the Solana transaction.

Usage:
    # Amazon product (by ASIN)
    python buy_and_sign.py --asin B0CXYZ1234 --email buyer@example.com --wallet 7xKXtg... --private-key 5abc... --address "John Doe,123 Main St,New York,NY,10001,US"

    # Shopify product (requires URL + variant)
    python buy_and_sign.py --url "https://store.com/products/item" --variant 41913945718867 --email buyer@example.com --wallet 7xKXtg... --private-key 5abc... --address "..."

Address format: "Name,Line1,City,State,PostalCode,Country[,Line2][,Phone]"

Required packages:
    pip install solana solders base58
"""

import argparse
import json
import sys
import urllib.request
import urllib.error

BASE_URL = "https://api.purch.xyz"


def parse_address(address_str: str) -> dict:
    """Parse address string into dict."""
    parts = address_str.split(",")
    if len(parts) < 6:
        raise ValueError("Address must have at least: Name,Line1,City,State,PostalCode,Country")

    address = {
        "name": parts[0].strip(),
        "line1": parts[1].strip(),
        "city": parts[2].strip(),
        "state": parts[3].strip(),
        "postalCode": parts[4].strip(),
        "country": parts[5].strip(),
    }

    if len(parts) > 6:
        address["line2"] = parts[6].strip()
    if len(parts) > 7:
        address["phone"] = parts[7].strip()

    return address


def create_order(
    email: str,
    wallet_address: str,
    shipping_address: dict,
    asin: str | None = None,
    product_url: str | None = None,
    variant_id: str | None = None,
) -> dict:
    """Create a purchase order."""
    payload = {
        "email": email,
        "walletAddress": wallet_address,
        "shippingAddress": shipping_address,
    }

    if asin:
        payload["asin"] = asin
    if product_url:
        payload["productUrl"] = product_url
    if variant_id:
        payload["variantId"] = variant_id

    data = json.dumps(payload).encode("utf-8")

    try:
        req = urllib.request.Request(
            f"{BASE_URL}/buy",
            data=data,
            headers={"Content-Type": "application/json"},
            method="POST"
        )
        with urllib.request.urlopen(req, timeout=60) as response:
            return json.loads(response.read().decode())
    except urllib.error.HTTPError as e:
        error_body = e.read().decode() if e.fp else ""
        return {"error": f"HTTP {e.code}: {error_body}"}
    except Exception as e:
        return {"error": str(e)}


def sign_and_send_transaction(
    serialized_tx: str,
    private_key: str,
    rpc_url: str = "https://api.mainnet-beta.solana.com"
) -> dict:
    """Sign and submit a Solana transaction."""
    try:
        import base58
        from solders.keypair import Keypair
        from solders.transaction import VersionedTransaction
        from solana.rpc.api import Client
        from solana.rpc.commitment import Confirmed
    except ImportError:
        return {"error": "Missing packages. Run: pip install solana solders base58"}

    # Decode private key
    try:
        key_bytes = base58.b58decode(private_key)
        keypair = Keypair.from_bytes(key_bytes)
    except Exception as e:
        return {"error": f"Invalid private key: {e}"}

    # Decode transaction
    try:
        tx_bytes = base58.b58decode(serialized_tx)
        transaction = VersionedTransaction.from_bytes(tx_bytes)
    except Exception as e:
        return {"error": f"Invalid transaction: {e}"}

    # Sign
    try:
        transaction.sign([keypair])
    except Exception as e:
        return {"error": f"Failed to sign: {e}"}

    # Send
    client = Client(rpc_url)

    try:
        result = client.send_transaction(transaction)
        signature = str(result.value)
        client.confirm_transaction(signature, commitment=Confirmed)

        return {
            "success": True,
            "signature": signature,
            "explorer_url": f"https://solscan.io/tx/{signature}"
        }
    except Exception as e:
        return {"error": f"Transaction failed: {e}"}


def main():
    parser = argparse.ArgumentParser(description="Create order and sign transaction")
    parser.add_argument("--asin", help="Amazon ASIN")
    parser.add_argument("--url", help="Product URL (Amazon or Shopify)")
    parser.add_argument("--variant", help="Variant ID (required for Shopify)")
    parser.add_argument("--email", required=True, help="Buyer email")
    parser.add_argument("--wallet", required=True, help="Solana wallet address")
    parser.add_argument("--private-key", required=True, help="Base58 private key for signing")
    parser.add_argument("--address", required=True, help="Shipping address")
    parser.add_argument("--rpc-url", default="https://api.mainnet-beta.solana.com", help="Solana RPC URL")
    parser.add_argument("--json", action="store_true", help="Output raw JSON")

    args = parser.parse_args()

    if not args.asin and not args.url:
        print("❌ Error: Either --asin or --url is required")
        sys.exit(1)

    try:
        shipping_address = parse_address(args.address)
    except ValueError as e:
        print(f"❌ Error: {e}")
        sys.exit(1)

    # Step 1: Create order
    print("📦 Creating order...")
    order_result = create_order(
        email=args.email,
        wallet_address=args.wallet,
        shipping_address=shipping_address,
        asin=args.asin,
        product_url=args.url,
        variant_id=args.variant,
    )

    if "error" in order_result:
        if args.json:
            print(json.dumps(order_result, indent=2))
        else:
            print(f"❌ Order creation failed: {order_result['error']}")
        sys.exit(1)

    order_id = order_result.get("orderId")
    serialized_tx = order_result.get("serializedTransaction")
    product = order_result.get("product", {})
    total_price = order_result.get("totalPrice", {})

    print(f"✅ Order created: {order_id}")
    print(f"   Product: {product.get('title', 'N/A')}")
    print(f"   Total: {total_price.get('amount', 'N/A')} {total_price.get('currency', 'USDC').upper()}")
    print()

    # Step 2: Sign and submit transaction
    print("🔐 Signing and submitting transaction...")
    tx_result = sign_and_send_transaction(serialized_tx, args.private_key, args.rpc_url)

    if args.json:
        print(json.dumps({"order": order_result, "transaction": tx_result}, indent=2))
        return

    if "error" in tx_result:
        print(f"❌ Transaction failed: {tx_result['error']}")
        print()
        print(f"🌐 You can complete payment via browser: {order_result.get('checkoutUrl')}")
        sys.exit(1)

    print("✅ Payment complete!")
    print(f"   Signature: {tx_result['signature']}")
    print(f"   Explorer:  {tx_result['explorer_url']}")


if __name__ == "__main__":
    main()

File: .skills/openclaw-skills/skills/purch-agent/agentic-commerce/scripts/buy_and_sign.ts
#!/usr/bin/env bun
/**
 * Create a purchase order AND sign/submit the Solana transaction.
 *
 * Usage:
 *   # Amazon product (by ASIN)
 *   bun run buy_and_sign.ts --asin B0CXYZ1234 --email buyer@example.com --wallet 7xKXtg... --private-key 5abc... --address "John Doe,123 Main St,New York,NY,10001,US"
 *
 *   # Shopify product (requires URL + variant)
 *   bun run buy_and_sign.ts --url "https://store.com/products/item" --variant 41913945718867 --email buyer@example.com --wallet 7xKXtg... --private-key 5abc... --address "..."
 *
 * Required packages:
 *   bun add @solana/web3.js bs58
 */

import { Connection, Keypair, VersionedTransaction, clusterApiUrl } from "@solana/web3.js";
import bs58 from "bs58";

const BASE_URL = "https://api.purch.xyz";

interface ShippingAddress {
  name: string;
  line1: string;
  line2?: string;
  city: string;
  state: string;
  postalCode: string;
  country: string;
  phone?: string;
}

interface OrderResponse {
  orderId: string;
  status: string;
  serializedTransaction: string;
  product: {
    title: string;
    imageUrl?: string;
    price: { amount: string; currency: string };
  };
  totalPrice: { amount: string; currency: string };
  checkoutUrl: string;
}

function parseAddress(addressStr: string): ShippingAddress {
  const parts = addressStr.split(",").map((p) => p.trim());
  if (parts.length < 6) {
    throw new Error("Address must have at least: Name,Line1,City,State,PostalCode,Country");
  }

  const address: ShippingAddress = {
    name: parts[0],
    line1: parts[1],
    city: parts[2],
    state: parts[3],
    postalCode: parts[4],
    country: parts[5],
  };

  if (parts[6]) address.line2 = parts[6];
  if (parts[7]) address.phone = parts[7];

  return address;
}

async function createOrder(params: {
  email: string;
  walletAddress: string;
  shippingAddress: ShippingAddress;
  asin?: string;
  productUrl?: string;
  variantId?: string;
}): Promise<OrderResponse> {
  const response = await fetch(`${BASE_URL}/buy`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(params),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`HTTP ${response.status}: ${error}`);
  }

  return response.json();
}

async function signAndSendTransaction(
  serializedTx: string,
  privateKey: string,
  rpcUrl: string = clusterApiUrl("mainnet-beta")
): Promise<{ success: boolean; signature?: string; explorerUrl?: string; error?: string }> {
  // Decode private key
  let keypair: Keypair;
  try {
    const keyBytes = bs58.decode(privateKey);
    keypair = Keypair.fromSecretKey(keyBytes);
  } catch (e) {
    return { success: false, error: `Invalid private key: ${e}` };
  }

  // Decode transaction
  let transaction: VersionedTransaction;
  try {
    const txBytes = bs58.decode(serializedTx);
    transaction = VersionedTransaction.deserialize(txBytes);
  } catch (e) {
    return { success: false, error: `Invalid transaction: ${e}` };
  }

  // Sign
  try {
    transaction.sign([keypair]);
  } catch (e) {
    return { success: false, error: `Failed to sign: ${e}` };
  }

  // Send
  const connection = new Connection(rpcUrl, "confirmed");

  try {
    const signature = await connection.sendRawTransaction(transaction.serialize(), {
      skipPreflight: false,
    });
    await connection.confirmTransaction(signature, "confirmed");

    return {
      success: true,
      signature,
      explorerUrl: `https://solscan.io/tx/${signature}`,
    };
  } catch (e) {
    return { success: false, error: `Transaction failed: ${e}` };
  }
}

function parseArgs(args: string[]): {
  asin?: string;
  url?: string;
  variant?: string;
  email?: string;
  wallet?: string;
  privateKey?: string;
  address?: string;
  rpcUrl?: string;
  json?: boolean;
} {
  const result: ReturnType<typeof parseArgs> = {};
  let i = 0;

  while (i < args.length) {
    const arg = args[i];
    if (arg === "--asin" && args[i + 1]) result.asin = args[++i];
    else if (arg === "--url" && args[i + 1]) result.url = args[++i];
    else if (arg === "--variant" && args[i + 1]) result.variant = args[++i];
    else if (arg === "--email" && args[i + 1]) result.email = args[++i];
    else if (arg === "--wallet" && args[i + 1]) result.wallet = args[++i];
    else if (arg === "--private-key" && args[i + 1]) result.privateKey = args[++i];
    else if (arg === "--address" && args[i + 1]) result.address = args[++i];
    else if (arg === "--rpc-url" && args[i + 1]) result.rpcUrl = args[++i];
    else if (arg === "--json") result.json = true;
    i++;
  }

  return result;
}

async function main() {
  const args = process.argv.slice(2);

  if (args.length === 0 || args[0] === "--help") {
    console.log(`
Usage: bun run buy_and_sign.ts [options]

Options:
  --asin <asin>           Amazon ASIN
  --url <url>             Product URL (Amazon or Shopify)
  --variant <id>          Variant ID (required for Shopify)
  --email <email>         Buyer email (required)
  --wallet <address>      Solana wallet address (required)
  --private-key <key>     Base58 private key for signing (required)
  --address <addr>        Shipping address (required)
  --rpc-url <url>         Solana RPC URL (default: mainnet-beta)
  --json                  Output raw JSON

Address format: "Name,Line1,City,State,PostalCode,Country[,Line2][,Phone]"
    `);
    process.exit(0);
  }

  const options = parseArgs(args);

  if (!options.asin && !options.url) {
    console.error("❌ Error: Either --asin or --url is required");
    process.exit(1);
  }
  if (!options.email || !options.wallet || !options.privateKey || !options.address) {
    console.error("❌ Error: --email, --wallet, --private-key, and --address are required");
    process.exit(1);
  }

  let shippingAddress: ShippingAddress;
  try {
    shippingAddress = parseAddress(options.address);
  } catch (e) {
    console.error(`❌ Error: ${e}`);
    process.exit(1);
  }

  try {
    // Step 1: Create order
    console.log("📦 Creating order...");
    const orderResult = await createOrder({
      email: options.email,
      walletAddress: options.wallet,
      shippingAddress,
      asin: options.asin,
      productUrl: options.url,
      variantId: options.variant,
    });

    console.log(`✅ Order created: ${orderResult.orderId}`);
    console.log(`   Product: ${orderResult.product?.title ?? "N/A"}`);
    console.log(`   Total: ${orderResult.totalPrice?.amount ?? "N/A"} ${(orderResult.totalPrice?.currency ?? "USDC").toUpperCase()}`);
    console.log();

    // Step 2: Sign and submit transaction
    console.log("🔐 Signing and submitting transaction...");
    const txResult = await signAndSendTransaction(
      orderResult.serializedTransaction,
      options.privateKey,
      options.rpcUrl || clusterApiUrl("mainnet-beta")
    );

    if (options.json) {
      console.log(JSON.stringify({ order: orderResult, transaction: txResult }, null, 2));
      return;
    }

    if (!txResult.success) {
      console.error(`❌ Transaction failed: ${txResult.error}`);
      console.log();
      console.log(`🌐 You can complete payment via browser: ${orderResult.checkoutUrl}`);
      process.exit(1);
    }

    console.log("✅ Payment complete!");
    console.log(`   Signature: ${txResult.signature}`);
    console.log(`   Explorer:  ${txResult.explorerUrl}`);
  } catch (error) {
    console.error(`❌ Error: ${error}`);
    process.exit(1);
  }
}

main();
